### DISCUSSION TITLE
FindClusters is taking too long for ~700K cells

### DISCUSSION THREAD

**Discussion by buutrg** [satijalab/seurat#8388] [Q&A]
Hi Seurat developers, I am using add Cluster on a 700K cells dataset, and it froze after Running Louvain algorithm for 12 hours. I have tried using default parameters with resolution = 1.5 and 1.8

```
Computing nearest neighbor graph
Computing SNN
Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 694117
Number of edges: 8990440

Running Louvain algorithm...
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Maximum modularity in 10 random starts: 0.9372
Number of communities: 11279
Elapsed time: 249 seconds
```

Is this a normal behavior? Do you know how to solve this? 

Thank you!

**Comment by sunhang526** [comment]
I have the similar problem now, have you solved it? please

**Comment by lizchcase** [comment]
Not sure if you already found your way through this but 700k is a ton of cells! This is great experimentally but will require a lot of computational resources. The amount of time it would take depends a lot on the specs of the computer/cluster you're running on. Generally, running with more nodes will make this go faster, and you will likely require quite a bit of memory/RAM

**Comment by xinglei01** [comment]
"Hello developers, I have a question. When the number of cells reaches over a million, the running speed of FindClusters with default parameters is too slow. Can I set singleton=F? What is the essential purpose of this parameter?"

**Comment by baike687** [comment]
I also have the same problem for about 500 000 cells, and It stays there for so long. Is it normal?
```r
Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 539830
Number of edges: 9682376

Running Louvain algorithm...
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Maximum modularity in 10 random starts: 0.9351
Number of communities: 2449
Elapsed time: 761 seconds
```

**Comment by HeesooSong** [comment]
Has anyone solved this issue? I'm stuck with 500k cells as well, but with only 950 genes...

**Comment by EduardGhemes-ICR** [comment]
Same problem here too, the algorithm is running too slow:
```
Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 977640
Number of edges: 26353598

Running Louvain algorithm...
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Maximum modularity in 10 random starts: 0.3874
Number of communities: 3174
Elapsed time: 584 seconds
```

**Comment by khadijah-mazhar** [comment]
I encountered the same issue with a dataset of 18K genes and 1 million cells. The process hanged for over 24 hours seemingly after completing the Louvain algorithm:

Running Louvain algorithm...
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Maximum modularity in 10 random starts: 0.9311
Number of communities: 7324
Elapsed time: 786 seconds

I killed the process and am not sure how long it would have taken to complete. I tried with both resolution = 1 and 0.8 (jn separate runs). 

I ended up using the sketch integration method: https://satijalab.org/seurat/articles/parsebio_sketch_integration. I ran FindClusters on 400K cells and then label transferred to the whole obj.
